[CT-472] feat: Retrying method for acquiring connection handles

tomasfarias opened the pull request:
resolves #5022 

<!---
  Include the number of the issue addressed by this PR above if applicable.
  PRs for code changes without an associated issue *will not be merged*.
  See CONTRIBUTING.md for more information.
-->

### Description

<!---
  Describe the Pull Request here. Add any references and info to help reviewers
  understand your changes. Include any tradeoffs you considered.
-->

Some adapters implement retrying functionality when attempting to open a connection. For example, see [dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/connections.py#L264=).

However, other adapters were missing this functionality, perhaps most notably dbt-postgres.

Although we could have just added a similar loop as the one in dbt-snowflake to dbt-postgres, this PR attempts to implement a common method `set_connection_handle` in `BaseConnectionManager` that any adapter may use when subclassing  `BaseConnectionManager`, in an attempt to provide a method and/or interface that all adapters can use, to avoid duplication.

The new method works by looping `retry_limit + 1` times. In each loop iteration we attempt to call `Connection.credentials.acquire_handle`. This latest method should be implemented by every adapter developer that wishes to utilize the retry functionality when subclassing `Credentials`, as it's heavily dependent on each adapter. For example, for dbt-postgres, the method would call `psycopg2.connect` with a given set of credentials.

On every loop, we check for exceptions we can handle, as specified in a `exception_handlers` dictionary. Upon encountering one, we log, call any additional function included in the dictionary, and retry. This allows other adapters, like dbt-bigquery, to call functions upon retrying specific exceptions (see sample commit in "In other adapters" section). Upon encountering an unexpected exception, we immediately break from the loop, set the connection state to FAIL, and raise a `FailedToConnectException` exception. However, if `retry_all` is set to `True`, we will be retrying on every exception regardless if it was unexpected. 

Moreover, this PR refactors dbt-postgres to use `set_connection_handle`, as well as adds some new `PostgresCredentials` attributes to customize the retry behavior. Most notably, we provide a `retry_on_errors` attribute with a default value of `"OperationalError"`. This attribute takes either a [PostgreSQL error code](https://www.postgresql.org/docs/current/errcodes-appendix.html) or a [psycopg2 exception class string](https://www.psycopg.org/docs/module.html#exceptions), upon which being raised during connecting a retry occurs. I chose `"OperationalError"` as a default value for two reasons:
* Connection errors, in particular timeouts, raise exceptions of this class.
* Timeouts specifically do not contain a PostgreSQL error code (`pgcode is None`), so a class name must be used instead.

#### Alternatives

An early implementation of `set_connection_handle` took a `Callable` argument instead of calling `Connection.credentials.acquire_handle`, and it was up to each subclass of `BaseConnectionManager` to provide this callable during `open`. However, given that the `Credentials` are taken from the `Connection` object itself, it seemed redundant to both pass the connection object and a callable that takes a credentials object when the connection already has the credentials in an attribute. Nonetheless, I added an optional `credentials` argument that overrides the `Connection.credentials`.

Another idea I had was to ignore `BaseConnectionManager` entirely, and implement the retrying functionality in `Credentials`. Then, just have each subclass implementation of `BaseConnectionManager.open` call `connection.credentials.acquire_handle()` (or a `connections.acquire_handle()` that simply passes the call). However, given that both dbt-snowflake and dbt-bigquery implement retrying in its subclasses of `BaseConnectionManager`, I discarded this idea. It goes without saying that if more context is shared  I'm more than willing to go with this alternative approach.

A second idea I had was to simply refactor the utility function [`_connection_exception_retry`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/utils.py#L608=). I opted not to go with this simply because adapters that have implemented retries did so without using it. That being said, this is something we could reconsider, and at the end of the day I do lack a lot of context to make this call.

#### In other adapters

Additionally to this PR, I've also pushed two commits to my forks of dbt-snowflake and dbt-bigquery to show how they may be refactored to use `set_connection_handle`, instead of their current implementations. I will keep these two commits up to date with any changes that we make here, and I will be more than glad to eventually open PRs for those two as well. That being said, I'm not familiar with dbt-snowflake and dbt-bigquery, so some input is required to set, for example, default retry-able errors, and credential attributes names (as they can conflict across adapters).

dbt-BigQuery commit: https://github.com/tomasfarias/dbt-bigquery/commit/488bd817f7b4350e465df403a6f2415bf4b02b04

dbt-Snowflake commit: https://github.com/tomasfarias/dbt-snowflake/commit/b97894ae85327aae224bac7c803d4c16174edc75

I noticed that dbt-bigquery [does some operations](https://github.com/dbt-labs/dbt-bigquery/blob/main/dbt/adapters/bigquery/connections.py#L308=) when it's choosing which credentials to use for connecting that are implemented in `BigQueryConnectionManager`. It appears to me that we could still implement a `acquire_handle` method in `BigQueryCredentials`, but this implementation will need reviewing, specially as I don't have experience with bigquery.

#### Testing

Naturally, I wanted to add testing for this new method, however I couldn't find any unit testing of `BaseConnectionManager` or `PostgresConnectionManager`, so I simply dropped a new file where I thought would make sense. I'm more than open to refactor this. Testing coverage may be extended if deemed necessary.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

nathaniel-may commented on test/unit/test_adapter_connection_manager.py:
```diff
@@ -0,0 +1,259 @@
+        assert conn.state == "fail"
+        assert conn.handle is None
+
+    def test_set_connection_handle_passes_handled(self):
```
This test looks good! Clear what it tests, and it does it pretty simply.

nathaniel-may reviewed (changes requested):
Thanks, @tomasfarias for this thoughtful and complex change. I definitely appreciate the detailed PR description and research you did to solve this very important issue to dbt users. I'm motivated to get this right, so I've left some extensive comments that I hope aren't too overwhelming. 

First, let's discuss the architecture of this solution at a high level and make sure we're on the same page: From what I understand, it looks like users are allowed to configure which exceptions are retryable in the `profiles.yml` of their projects. This gets picked up after parsing in the `Credentials` class which allows us to call `BaseConnectionManager.set_connection_handle` in the `PostgresConnectionManager.open` method instead of directly calling `psycopg2.connect` which does not have built-in connection retries. `BaseConnectionManager.set_connection_handle` is a while loop which calls specific catch functions for each retryable exception that we encounter and retries `acquire_handle`.

What I like about this approach is that it provides the option for current adapter maintainers to modify their `open` implementations to use `set_connection_handle` instead, but their existing implementation will work just fine until they chose to update it.

My instinct here is that we can simplify this solution to make it easier to follow and to implement for adapter maintainers. Here are my suggestions:
1. The exceptions that are retryable generally do not change between dbt projects. Instead of setting them at the dbt project level, I think they should be set at the adapter level to specify which errors for that warehouse are retryable. If there's a good use-case for them to be overriden by certain projects, I'd love to see some examples. There might be a better way to provide that kind of override if it really is necessary.
1. With regards to the exception handler functions, I want to distinguish between retrying errors and recovering from errors. The main goal here is to retry the errors, not necessarily recover from them. Additionally, your implementation for the postgres adapter doesn't actually use these recovery functions. So in the interest of simplifying this implementation I want to consider using an inclusive list instead of a dictionary. If you disagree and think the recovery functions are necessary I would love to see some examples of situations we would need them.
1. If `acquire_handle` lived in the `PostgresConnectionManager` class instead of the `Credentials`, it would simplify the implementation of `set_connection_handle` because we wouldn't need to take the optional credentials param and check it.
1. An alternative to `retry_all` might be to allow adapter maintainers to provide an exclusive list to `BaseConnectionManager.set_connection_handle` so that you could either call it with an inclusive list of exceptions to retry, or and exclusive list of exceptions not to retry.
1. We might want to explore some name alternatives for clarity and so we don't overuse the word "handle." Let's come up with an alternative for `set_connection_handle` that makes it more clear that this function is a retry wrapper.

I've left several comments in the code itself with "blocker" or "nit" to distinguish between what I need to change or discuss before merging and what is a nice-to-have if you agree with my suggestion.

This is a really important change that I want to make sure we get right. I really appreciate all the work you put into this PR and I'm looking forward to hammering out the details with you and getting it into dbt!

Special thanks to @dataders for walking through this PR with me.

tomasfarias commented:
Hey @nathaniel-may.

Thanks for the detailed feedback, I'll be addressing all your comments. Moreover, here's some questions and comments about your feedback:

> What I like about this approach is that it provides the option for current adapter maintainers to modify their open implementations to use set_connection_handle instead, but their existing implementation will work just fine until they chose to update it.

This was the idea, I didn't want to break existing adapters but instead provide an opt-in solution.

> The exceptions that are retryable generally do not change between dbt projects. Instead of setting them at the dbt project level, I think they should be set at the adapter level to specify which errors for that warehouse are retryable.

This is a good observation and I agree with it: I can't imagine different retryable exceptions for different projects with the same adapter. The idea for using `profiles.yml` for exposing retry parameters came from [dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/connections.py#L64=).

That being said, does this mean we should not allow users to set errors to retry? Where would the errors to retry be set? As an extra attribute of `AdapterConfig` (e.g. [here](https://github.com/dbt-labs/dbt-core/blob/main/plugins/postgres/dbt/adapters/postgres/impl.py#L54=) for the postgres adapter)?

> With regards to the exception handler functions, I want to distinguish between retrying errors and recovering from errors. The main goal here is to retry the errors, not necessarily recover from them. Additionally, your implementation for the postgres adapter doesn't actually use these recovery functions. So in the interest of simplifying this implementation I want to consider using an inclusive list instead of a dictionary.

The main reason I opted for a dictionary and the inclusion of recovery functions was to allow dbt-bigquery to use them. See [here](https://github.com/dbt-labs/dbt-bigquery/blob/main/dbt/adapters/bigquery/connections.py#L333=) where they recover from a specific `except google.auth.exceptions.DefaultCredentialsError`. [My fork](https://github.com/tomasfarias/dbt-bigquery/blob/main/dbt/adapters/bigquery/connections.py#L334=) shows how dbt-bigquery could resolve this with the new `set_connection_handle`.

That being said, indeed dbt-postgres does not use any recovery functions, and if the dbt-bigquery scenario is deemed too specific, then a list would just work too. dbt-bigquery can keep their try/except in place instead, and our method can re-`raise` for them.

> An alternative to retry_all might be to allow adapter maintainers to provide an exclusive list to BaseConnectionManager.set_connection_handle.

The `retry_all` idea was also taken from [dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/connections.py#L65=). That being said, I'm kind of neutral between both, so we can try an exclusive list to see how it looks. The questions I made before about where would this be configured on the first point still apply here.

> We might want to explore some name alternatives for clarity and so we don't overuse the word "handle." Let's come up with an alternative for set_connection_handle that makes it more clear that this function is a retry wrapper.

Personally, I like `open`, however this method is currently part of the `BaseConnectionManager` interface and changing it would require all adapters to be refactored, so I tried to look for other names. Naming things is hard though, so here are a few ideas to inspire (in no particular order):

* `retry_connection` / `retry_connect`
* `connect`
* `open_connection`

tomasfarias commented on core/dbt/adapters/base/connections.py:
```diff
@@ -159,6 +162,131 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+                is_expected = any(issubclass(type(e), exc) for exc in expected_exceptions)
+                if is_expected:
+                    logger.debug(
+                        "Got an error when attempting to open a "
+                        "{conn_type} connection. Retrying due to "
+                        "either retry configuration set to true."
+                        "This was attempt number: {attempt} of "
+                        "{max_attempts}. "
+                        "Retrying in {timeout} "
+                        "seconds. Error: '{error}'".format(
+                            conn_type=cls.TYPE,
+                            attempt=attempt,
+                            max_attempts=max_attempts,
+                            timeout=timeout,
+                            error=e,
+                        )
```
:+1: I also like f-strings. That being said, personally I would have gone with %s-style formatting (or `{}`) to allow `logging` to defer message formatting as the [Python docs describe](https://docs.python.org/3/howto/logging.html#optimization). However, this was not possible as we are using [`AdapterLogger`](https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/events/adapter_endpoint.py#L12=) instead of plain `logging`.

nathaniel-may replied:
Oh you can still do that with the AdapterLogger. [Here's an example from the postgres adapter.](https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/plugins/postgres/dbt/adapters/postgres/connections.py#L78).

tomasfarias commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -22,6 +22,7 @@ class PostgresCredentials(Credentials):
     user: str
     port: Port
     password: str  # on postgres the password is mandatory
+    connect_retries: int = 3
```
Yes, you are correct.

This idea was taken from dbt-snowflake which already implements it's own version of retrying. That being said, I'm not a fan of the confusion between retry options and connection options: for example, `connect_timeout` is a Postgres connection option, yet in dbt-snowflake it's a retry option, which is why here I'm using `retry_timeout` instead, but this means different adapters have different configuration names for what it's essentially the same thing.

nathaniel-may replied:
@jtcohen6 do you have any opinions here?

tomasfarias replied:
Personally, I'm a fan of extensive configurations with sensible defaults, but defining said defaults can be hard, and in the mean-time the user experience would suffer. So, I do see the argument for simplicity, and I thank you for your insight.

The question that is still open for me is where would the `number_of_retries` configuration option be exposed. I think there are 3 options:
1. As a user configuration (similar to `fail_fast`, `use_colors`, or other [flags](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/flags.py)).
2. As a top level profile option (similar to `target`).
3. In each individual target, which means in each adapters credentials (next to other connection options like `password`, that's how it's done in this PR currently).

Given that each target of a given adapter will have the same retrying behavior, I think the 3rd option would be missed a lot or would lead to a lot of duplication.

I haven't see top level profile options used for anything but the default target and outputs, so I'd think 2nd option would also be out.

And for these reasons I'm leaning towards 1. But do you have any thoughts on this?

tomasfarias replied:
Yeah, `retries` definitely belongs in with other connection options, I do agree with that. From a user perspective, I still think it would be nice to be able to set a global `retries`, but that can perhaps be left for the future as I do agree that it doesn't go well with other CLI flags.

I'll try this out in a follow-up commit. Thanks :+1:

tomasfarias commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -31,6 +32,13 @@ class PostgresCredentials(Credentials):
     application_name: Optional[str] = "dbt"
+    retry_all: bool = False
+    retry_timeout: int = 5
+    retry_on_errors: List[str] = field(
+        default_factory=lambda: [
+            "OperationalError"  # Covers all common connection errors, like timeouts.
+        ]
+    )
```
Setting a mutable type default argument (like a list) raises an error with dataclasses. The [Python docs describe the issue](https://docs.python.org/3/library/dataclasses.html#mutable-default-values) and recommend using `default_factory`. The basic gist is that with multiple instances of the class, both instances share the same underlying list, which can lead to bugs.

tomasfarias replied:
To add to the explanation, this is the Exception raised if you attempt to use a `list` as a default argument:
```
ValueError: mutable default <class 'list'> for field retry_on_errors is not allowed: use default_factory
```

tomasfarias commented on core/dbt/adapters/base/connections.py:
```diff
@@ -159,6 +162,131 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+        expected_exceptions = exception_handlers.keys() if exception_handlers else ()
+        exception_handlers = exception_handlers if exception_handlers else {}
+        max_attempts = 1 + retry_limit
+        acquire_handle = (
+            credentials.acquire_handle
+            if credentials is not None
+            else connection.credentials.acquire_handle
```
How exactly would this work? If we move `acquire_handle` to `PostgresConnectionManager` we still need to pass the credentials to `acquire_handle`, but that's not what you are saying as `connection.credentials.acquire_handle` is still using `acquire_handle` from the `Credentials` class.

tomasfarias commented:
My latest commits should address most requested comments. I've asked for clarification on how to refactor `acquire_handle` so that is still open.

Just to summarize why I don't understand the `acquire_handle` request:

Moving `acquire_handle` to `PostgresConnectionManager` would require obtaining the credentials from somewhere else, which can be achieved by calling `cls.acquire_handle(connection.credentials)` or `cls.acquire_handle(connection)` and accessing the attribute in the body of `acquire_handle`.

However, a comment mentions `acquire_handle = connection.credentials.acquire_handle` should be used, but this is still accessing `acquire_handle` in the Credentials object, which would no longer exist once it's moved to `PostgresConnectionManager`.

nathaniel-may commented on test/unit/test_adapter_connection_manager.py:
```diff
@@ -0,0 +1,259 @@
+from dbt.events import AdapterLogger
+
+
+class BaseConnectionManagerTest(unittest.TestCase):
```
Thanks for adding all the comments! They look great.

jtcohen6 commented:
@tomasfarias I think the motivation behind `retry_all`, on `dbt-spark` and `dbt-snowflake`, is for adapters/drivers where we lack confidence in the set of exceptions which we believe to be intermittent/transient, and therefore retryable. I think we _should_ aspire to have that set of exceptions on all adapters, and shouldn't need to expose a distinction to end users between `retry_the_ones_we_think_are_maybe_retryable` and `retry_all` (better safe than sorry).

So: I really appreciate the work you've done to research the existing implementations, and aim for consistency / uniformity. I do not think `retry_all` must be part of our final design here.

(`retry_all: True` is also the implicit behavior in `dbt-bigquery`, and it really shouldn't be!)

nathaniel-may commented:
Hey, @tomasfarias! We are targeting getting this PR into dbt 1.2.0rc2 which would mean we need to get it merged before the end of business on Tuesday 07/19. I'm going to finish answering your questions today so we can work out the changes, but if you don't have the time and would rather me take over the PR I can also do that.

tomasfarias commented:
> Hey, @tomasfarias! We are targeting getting this PR into dbt 1.2.0rc2 which would mean we need to get it merged before the end of business on Tuesday 07/19. I'm going to finish answering your questions today so we can work out the changes, but if you don't have the time and would rather me take over the PR I can also do that.

That's great to hear @nathaniel-may, thank you. I'll have plenty of time this week so I'll be happy to make a push for a Tuesday 07/19 release. Let me go over your recent comments.

nathaniel-may commented:
That's great to hear! Thanks again for working with us on this one.

I played around with the change I wanted for `acquire_handle` and it turns out that just doesn't work because I can't access instance values in a class method and it wasn't clear to me I'd run into that till I tried to do it myself. I came up with what I think is a simpler alternative and I want to run by you to see what you think:

My goal here was to treat `set_connection_handle` as the thinnest of retry wrappers possible. To do this, I moved the `acquire_handle` method to just be a value inside `open` that you pass to `set_connection_handle`. I renamed a lot of things and changed the while loop to recursion to reduce the number of variables. (I know Python and recursion aren't friends but this retry loop should always be less than 10 for reasonable configurations.) Full disclaimer that I didn't modify all your test cases and run them on this code yet.

Here's how it would be used in `open`:

```python
class PostgresConnectionManager(SQLConnectionManager):
    # ... other methods ...
    
    @classmethod
    def open(cls, connection):
        # ... other open stuff ...
        
        # nested function definition
        def connect():
            handle = psycopg2.connect(
                dbname=credentials.database,
                user=credentials.user,
                host=credentials.host,
                password=credentials.password,
                port=credentials.port,
                connect_timeout=credentials.connect_timeout,
                **kwargs,
            )
            if self.role:
                handle.cursor().execute("set role {}".format(self.role))
            return handle

        return cls.retry_connection(
            connection,
            connect=connect,
            logger=logger,
            retryable_exceptions=[Exception],  # retries all TODO fix.
            retry_limit=3  # TODO expose this to user
        )
```

and here's the definition:

```python
class BaseConnectionManager(metaclass=abc.ABCMeta):
    # ... other methods ...
    
    # renamed from set_connection_handle
    @classmethod
    def retry_connection(
        cls,
        connection: Connection,
        connect: Callable[[], Connection],
        logger: AdapterLogger,
        retryable_exceptions: List[Type],
        retry_limit: int = 1,
        retry_timeout: int = 1
    ) -> Connection:
        """A retry function for Connection objects that manages Connection state
        and retryability of exceptions.
        # TODO full doc string
        """

        # sleep(x) throws if x is negative. throwing our own exception here for ease of use.
        if retry_timeout < 0:
            raise dbt.exceptions.FailedToConnectException("retry_timeout must be non-negative.")

        # this guards when called initially with a retry_limit < 0.
        # during normal execution we will catch when we run out of
        # retries in the except block.
        if retry_limit <= -1:
            connection.handle = None
            connection.state = ConnectionState.FAIL
            raise dbt.exceptions.FailedToConnectException("no connection attempts left.")

        # if we successfully connect, set state and return
        try:
            connection.handle = connect()
            connection.state = ConnectionState.OPEN
            return connection

        # if we catch an exception retry if it's retryable and we have retries left.
        except Exception as e:
            retryable = any(issubclass(type(e), exc) for exc in retryable_exceptions)
            if retryable:
                # no more retries left and we haven't returned a connection object yet.
                # throw the last encountered exception.
                if retry_limit <= 0:
                    connection.handle = None
                    connection.state = ConnectionState.FAIL
                    raise dbt.exceptions.FailedToConnectException(str(e))

                # this message only logs if we have retries left.
                # if not, it throws above
                logger.debug(
                    f"Got a retryable error when attempting to open a {cls.TYPE} connection.\n"
                    f"{retry_limit} attempts remaining. Retrying in {retry_timeout} seconds.\n"
                    f"Error:\n{e}"
                )

                # since we haven't exited the function yet
                # we should retry with one less retry_limit
                sleep(retry_timeout)
                return cls.retry_connection(
                    connection=connection,
                    connect=connect,
                    logger=logger,
                    retry_limit=retry_limit-1,
                    retry_timeout=retry_timeout,
                    retryable_exceptions=retryable_exceptions
                )
            else:
                logger.warning(
                    f"Got an error that cannot be retried when attempting to open a {cls.TYPE} connection."
                )
                connection.handle = None
                connection.state = ConnectionState.FAIL
                raise dbt.exceptions.FailedToConnectException(str(e))
```

nathaniel-may commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -54,6 +62,22 @@ def _connection_keys(self):
             "sslmode",
         )
 
+    def acquire_handle(self, **kwargs):
+        """Acquire a handle for a Postgres connection with the given credentials."""
+        handle = psycopg2.connect(
+            dbname=self.database,
+            user=self.user,
+            host=self.host,
+            password=self.password,
+            port=self.port,
+            connect_timeout=self.connect_timeout,
+            **kwargs,
+        )
+
+        if self.role:
+            handle.cursor().execute("set role {}".format(self.role))
+        return handle
```
Turns out this doesn't work. I present a more feasible alternative in [this comment](https://github.com/dbt-labs/dbt-core/pull/5432#issuecomment-1183787447).

tomasfarias commented:
Thanks for your write-up @nathaniel-may, here are some of my opinions:

>  (I know Python and recursion aren't friends but this retry loop should always be less than 10 for reasonable configurations.)

Nit: We could do:

```python
import sys
...
        if retry_limit < 0 or retry_limit > sys.getrecursionlimit():
            connection.handle = None
            connection.state = ConnectionState.FAIL
            raise dbt.exceptions.FailedToConnectException(
                f"retry_limit cannot be negative or more than {sys.getrecursionlimit()}"
            )
```

To be extra safe...

> To do this, I moved the acquire_handle method to just be a value inside open that you pass to set_connection_handle

An initial idea I had worked exactly like this: `set_connection_handle` just took a callable like your `connect`. I believe the only difference is that I just defined the function in the global scope instead of nested, and took a `Credentials` object instead of making use of the credentials in the local scope.

Having the function in the global scope would allow adapters that build on top of the postgres adapter to re-use the function without having to re-write it in `open`. However, this is a bit of a stretch, and as far as I can see, it's not the case with dbt-redshift. So, overall, I don't feel strongly regarding scope.

This idea is a bit tricky to type, though. I've noticed you have done the following:

```python
connect: Callable[[], Connection],
```

I assume the return `Connection` type is a `dbt.contracts.connection.Connection` object, as it's also used for the second argument. However, every adapter that implements its own `connect` callable will return an object that is not a `dbt.contracts.connection.Connection` (for example, your `connect` function returns a `psycopg2.connection` object).

Some proper typing solution may require adapter maintainers to register the return types as virtual subclasses of a new `AdapterHandle` metaclass (for example, although not being included in the MRO means this serves no purpose besides being a reminder for maintainers).

This could be quite the burden, so I would just opt for a type alias (`AdapterHandle`, for example) and good documentation:

```python
from typing import Optional, Any

# Btw, the type hint of dbt.contracts.connection.Connection._handle is Optional[Any]
AdapterHandle = Any

class BaseConnectionManager(metaclass=abc.ABCMeta):
    # ... other methods ...
    
    # renamed from set_connection_handle
    @classmethod
    def retry_connection(
        cls,
        connection: Connection,
        connect: Callable[[], AdapterHandle],
```

Moving on, I've noticed that we have simplified things and only taking a single inclusive list of exceptions:
```python
        retryable_exceptions: List[Type],
```
I like this :+1:, honestly I'm not too happy with how I implemented the added complexity of handling both an inclusive and exclusive list.

On this TODO:

```python
retryable_exceptions=[Exception],  # retries all TODO fix.
```

I think `psycopg2.error.OperationalError` is a sensible default (it's implemented like this currently, and I've left a comment explaining what are we retrying on).

Somewhat nitpicky typing:

```python
retryable_exceptions: List[Type],
```

The only thing we do with `retryable_exceptions` is iterate over it, so I would be more lenient and take an `Iterable` instead of a `List`. This allows adapter maintainers to pass tuples, sets, or even generators, besides lists. I would be more strict with the type of the contents, though, as we only expect subclasses of `Exception` here. So: 

```python
retryable_exceptions: Iterable[Type[Exception]],
```

On retry_timeout:

```python
retry_timeout: int = 1
```

Are we discarding the current implementation that also allows a callable? For example, if adapter maintainers wish to pass a function that implements some exponential backoff algorithm. Also, `sleep` can take `float`s.

On a design aspect:

Given that we have simplified things via recursion and gotten rid of some optional arguments, why not get rid of the `if retryable` check too:

```python
        # Catches all retryable exceptions
        except tuple(retryable_exceptions) as e:
            if retry_limit <= 0:
                connection.handle = None
                connection.state = ConnectionState.FAIL
                raise dbt.exceptions.FailedToConnectException(str(e))

            # log, sleep, recursive call, etc...
        # Catches everything else
        except Exception as e:
            logger.warning(
                f"Got an error that cannot be retried when attempting to open a {cls.TYPE} connection."
            )
            connection.handle = None
            connection.state = ConnectionState.FAIL
            raise dbt.exceptions.FailedToConnectException(str(e))
```

I'll combine my comments + your implementation in my next commit so that we can work over it and hopefully finish this :+1:

nathaniel-may commented on core/dbt/adapters/base/connections.py:
```diff
@@ -159,6 +176,88 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+            connection.state = ConnectionState.OPEN
+            return connection
+
+        except tuple(retryable_exceptions) as e:
```
oh yeah this is much slicker.

nathaniel-may commented on core/dbt/contracts/connection.py:
```diff
@@ -178,6 +178,9 @@ def __post_serialize__(self, dct):
             )
         return dct
 
+    def acquire_handle(self, **kwargs):
+        raise NotImplementedError
```
blocker: I don't think we need this anymore right?

tomasfarias replied:
Ah, yeah, remnants from a previous implementation. I'll remove it.

nathaniel-may commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -131,23 +148,23 @@ def open(cls, connection):
-        return connection
+            return handle
+
+        retryable_exceptions = [
+            # OperationalError is subclassed by all psycopg2 Connection Exceptions
+            # and it's raised by generic connection errors without an error code.
+            psycopg2.errors.OperationalError,
+        ]
```
I think you're right that all the exceptions we want to retry are subclasses of `OperationalError`. However, this set includes invalid passwords and the disk being full. This is a smaller subset of exceptions I think we should retry on connection:

ConnectionException
SqlclientUnableToEstablishSqlconnection
ConnectionDoesNotExist
SqlserverRejectedEstablishmentOfSqlconnection
ConnectionFailure
TooManyConnections

Unfortunately the psychopg2 docs don't have much info on these exceptions nor do the postgres docs. I had to go to postgres source code to figure out what these exceptions do and of course, I could be wrong.

tomasfarias replied:
The problem is that psycopg2 raises `OperationalError`s on generic timeouts (i.e. when waiting for a response from the database upon attempting to connect exceeds `connect_timeout`). The exception message looks like:

This is because specific subclasses are implemented only for errors that have an associated SQLSTATE, or messages emitted by the database server. But as we haven't connected to the database yet, there is no message from the database server!

So we need to catch `OperationalError`. And if we are going to catch `OperationalError` we might as well exclude all other subclasses as they are already caught too.

tomasfarias replied:
There is an issue on the `psycopg2` GitHub about `OperationalError` being too generic: https://github.com/psycopg/psycopg2/issues/682

But unfortunately it only addressed the creation of individual subclasses for errors that have an associated PGCODE. Errors without a PGCODE (like generic timeouts) have PGCODE set to None, and it doesn't look it will get a subclass:

OP:
> In some cases, particularly around authentication, pgcode is None and so there is no alternative (that I have found) to parsing the error message.

Maintainer:
> No, we don't have plan to add other information other than what the server or the postgres client library provides us, only to make easier to access them.
> I feel your pain though. You should ask the postgres developer to add such a code. Note that those messages are susceptible of localization, apart from being more likely than a code to change from one version to the other.

tomasfarias replied:
In order to prove this is how timeouts work, I did the following:

1. I found a tool that can simulate a slow TCP proxy: https://github.com/Shopify/toxiproxy. Installed and ran a server.
2. I setup a local connection to a Redshift cluster: `localhost:5439`. Connection is the same I use for development, and works without issues.
3. I setup a proxy that redirects traffic to the local Redshift cluster using the toxiproxy-cli: `toxiproxy-cli create --listen localhost:65432 --upstream localhost:5439 redshift-proxy`
4. I ran the following Python Program (credentials omitted from dsn). Notice that `connect_timeout=1`:

```python
import psycopg2

try:
    conn = psycopg2.connect("host=localhost port=65432 connect_timeout=1")
except Exception as e:
    saved = e
```
Before adding latency this produces a proper connection:

```python
In [3]: conn
Out[3]: <connection object at 0x7f6349895a80; dsn: 'host=localhost port=65432 connect_timeout=1', closed: 0>
```

6. I added some latency: `toxiproxy-cli toxic add -t latency -a latency=10000 redshift-proxy`
7. I re-ran the same Python program as in 4:

```python
In [3]: conn
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 conn

NameError: name 'conn' is not defined

In [4]: str(saved)
Out[4]: 'connection to server at "localhost" (::1), port 65432 failed: Connection refused\n\tIs the server running on that host and accepting TCP/IP connections?\nconnection to server at "localhost" (127.0.0.1), port 65432 failed: timeout expired\n'

In [5]: assert saved.pgcode is None

In [6]: assert saved.pgerror is None

In [7]: saved.__class__
Out[7]: psycopg2.OperationalError
```

This time, due to the added latency being higher than 1, connecting fails. `psycopg2` raises a `psycopg2.OperationalError` with a `timeout expired` error message. This exception object *doesn't have a pgcode*.

nathaniel-may replied:
Well that's rather unfortunate. `OperationalError` it is then. I would add a comment with the link to that bug that's blocking us from being more specific in our retries.

Instead of a `List[Type[Exception]]` we could take a `Callable[[Type[Exception]], bool]` but that might be just be confusing overkill for the interface. I'm fine for the Postgres implementation just passing `[OperationalError]`.

nathaniel-may commented on test/unit/test_adapter_connection_manager.py:
```diff
@@ -0,0 +1,424 @@
+        This test uses a Connection populated with test PostgresCredentials values, and
+        expects the Connection.handle attribute to be set to True and it's state to
+        "open". Moreover, this must happen in the first attempt as no exception would
+        be raised for retrying. A mock acquire_handle is set to simulate a real connection
```
blocker: I think this comment needs to be updated.

dataders commented on core/dbt/adapters/base/connections.py:
```diff
@@ -34,6 +48,9 @@
 )
 from dbt import flags
 
+SleepTime = Union[int, float]  # As taken by time.sleep.
```
alternatively, is there even a use case for specifying seconds more specifically that integer granularity?

tomasfarias replied:
Not that I can think of. I opted for the union as that's what `time.sleep` can take, so we can leave it up to adapter maintainers to use whatever they want.

I would be okay with restricting the type hint though, as I don't think it's very critical either way.

nathaniel-may replied:
@dataders there's also not a use case to restrict the input values to just ints. since float is (roughly) a superset of int, I'd rather just choose the bigger type. The union is totally functional, it's just add a level of indirection I don't think is totally necessary.

nathaniel-may replied:
either option of `float` or `SleepTime = Union[int, float]` is fine with me.

nathaniel-may commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -131,23 +148,23 @@ def open(cls, connection):
+            connection,
+            connect=connect,
+            logger=logger,
+            retry_limit=credentials.retries,
```
RE: your conversation with @jtcohen6, is credentials the place we've decided to expose this value?

tomasfarias replied:
I proposed 3 potential places where we could expose the `retries` configuration: https://github.com/dbt-labs/dbt-core/pull/5432#discussion_r920320624

@jtcohen6 felt inclined towards going with the third option (and I agreed):

> 3. In each individual target, which means in each adapters credentials (next to other connection options like `password`, that's how it's done in this PR currently).

The linked discussion has more details about our discussion.

tomasfarias commented on core/dbt/adapters/base/connections.py:
```diff
@@ -159,6 +176,88 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+            depending on whether we successfully acquired a handle or not.
+        """
+        timeout = retry_timeout(retry_limit) if callable(retry_timeout) else retry_timeout
+        if timeout < 0:
```
I'll add the recursion limit check :+1: 

Given that `time.sleep` can take 0, I don't think we would need to check for equality in `timeout <= limit` as we are not preventing any exceptions.

tomasfarias commented on core/dbt/adapters/base/connections.py:
```diff
@@ -159,6 +176,88 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+        logger: AdapterLogger,
+        retryable_exceptions: Iterable[Type[Exception]],
+        retry_limit: int = 1,
+        retry_timeout: Union[Callable[[int], SleepTime], SleepTime] = 1,
```
The type signature is definitely simpler, and, although exp. backoff could work with some state that the caller has (as it knows the max. retries when calling `retry_connection`), it's not very intuitive. An increasing `int` feels right.

Personally, I lean towards having a `attempts: int = 0` + docs just because it's simpler to implement. Perhaps with a name to indicate that it's an internal counter: `attempt_counter: int = 0`, `internal_attempts: int = 0`, or an underscore `_attempts: int = 0`.

tomasfarias replied:
That being said, I'm not sure I like the removal of the option to just pass an int/float. Sure the type signature is simpler, but from a caller perspective, `lambda: 1` is not intuitive. Perhaps we could go with different arguments?

For example:

```
    @classmethod
    def retry_connection(
        cls,
        connection: Connection,
        connect: Callable[[], AdapterHandle],
        logger: AdapterLogger,
        retryable_exceptions: Iterable[Type[Exception]],
        retry_limit: int = 1,
        retry_timeout: SleepTime = 1,
        retry_timeout_func: Optional[Callable[[int], SleepTime]] = None,
    ) -> Connection:
        timeout = retry_timeout_func(...) if retry_timeout_func else retry_timeout
```

nathaniel-may replied:
for the sake of time, let's just add in `_attempts: int = 0`. Keeping the type a union of callable and sleeptime is fine.

tomasfarias replied:
It was added in latest commit.

nathaniel-may commented on plugins/postgres/dbt/adapters/postgres/connections.py:
```diff
@@ -54,6 +55,22 @@ def _connection_keys(self):
             "sslmode",
         )
 
+    def acquire_handle(self, **kwargs):
```
blocker: I think this needs to be removed.

nathaniel-may commented:
@tomasfarias yes we are very close! I added what input I think you were missing. If there's anything else you're waiting on that I might have missed please raise it again so I can address it.

tomasfarias commented:
@nathaniel-may Thanks for your input! I believe all comments should be addressed, but do let me know if there's anything still open.

nathaniel-may commented on .changes/unreleased/Features-20220715-035555.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: Features
+body: Support retrying Postgres (and Redshift!) connection attempts in the prescence
```
nit: better changelog entry since this change is larger in scope than this body suggests. Maybe something like "Add reusable function for retrying adapter connections. Adds connection retries for Postgres (and Redshift)."

tomasfarias replied:
Updated :+1:. Agree that the previous message didn't capture the scope.

nathaniel-may reviewed (approved):
I think we did it! This looks really great. Thanks for all the hard work you put in here. This will be so awesome to have in 1.2. I added one comment about improving the changelog entry, but I'm approving regardless.

dataders commented:
so stoked for this @tomasfarias -- thanks for your perseverance!

leahwicz commented:
Is the Snyk check the only thing holding this PR back at this point? I think if we close and re-open the PR, it will re-run the check which I'm guessing failed due to a Snyk outage since it says it could not complete analysis.

dataders commented:
> Is the Snyk check the only thing holding this PR back at this point? I think if we close and re-open the PR, it will re-run the check which I'm guessing failed due to a Snyk outage since it says it could not complete analysis.

that's my understanding!
